pkg/controllers/nodepool/status/controller.go

```diff
@@ -0,0 +1,134 @@
+*/
+
+package status
+
+import (
+	"context"
+	"fmt"
+
+	"go.uber.org/multierr"
+	"k8s.io/apimachinery/pkg/api/errors"
+
+	"sigs.k8s.io/karpenter/pkg/utils/nodepool"
+
+	"github.com/awslabs/operatorpkg/status"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"knative.dev/pkg/logging"
+
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+	"sigs.k8s.io/karpenter/pkg/operator/injection"
+
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"sigs.k8s.io/karpenter/pkg/apis/v1beta1"
+)
+
+// Controller for the resource
+type Controller struct {
+	kubeClient    client.Client
+	cloudProvider cloudprovider.CloudProvider
+}
+
+// NewController is a constructor
+func NewController(kubeClient client.Client, cloudProvider cloudprovider.CloudProvider) *Controller {
+	return &Controller{
+		kubeClient:    kubeClient,
+		cloudProvider: cloudProvider,
+	}
+}
+
+func (c *Controller) Reconcile(ctx context.Context, nodePool *v1beta1.NodePool) (reconcile.Result, error) {
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named("nodepool.status").With("nodepool", nodePool.Name))
+	ctx = injection.WithControllerName(ctx, "nodepool.status")
+	stored := nodePool.DeepCopy()
+
+	nodeClass := &unstructured.Unstructured{}
+	nodeClass.SetAPIVersion(nodePool.Spec.Template.Spec.NodeClassRef.APIVersion)
+	nodeClass.SetKind(nodePool.Spec.Template.Spec.NodeClassRef.Kind)
+	supportedNC := c.cloudProvider.GetSupportedNodeClasses()
+	if len(nodeClass.GetAPIVersion()) == 0 {
+		nodeClass.SetAPIVersion(fmt.Sprintf("%s/%s", supportedNC[0].Group, supportedNC[0].Version))
+	}
+	if len(nodeClass.GetKind()) == 0 {
+		nodeClass.SetKind(supportedNC[0].Kind)
+	}
+	if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: nodePool.Spec.Template.Spec.NodeClassRef.Name}, nodeClass); err != nil {
+		return reconcile.Result{}, fmt.Errorf("resolving nodeClass, %w", err)
+	}
+	ready, err := c.IsReady(status.ConditionReady, nodeClass)
+	if err != nil || !ready {
+		nodePool.StatusConditions().SetFalse(v1beta1.ConditionTypeNodeClassReady, "NodeClassNotReady", "unable to resolve underlying nodeClass")
+	}
+	if ready {
+		nodePool.StatusConditions().SetTrue(v1beta1.ConditionTypeNodeClassReady)
+	}
+	if !equality.Semantic.DeepEqual(stored, nodePool) {
+		if e := c.kubeClient.Status().Update(ctx, nodePool); e != nil {
+			err = multierr.Append(err, client.IgnoreNotFound(e))
+		}
+	}
+	if err != nil {
+		if errors.IsConflict(err) {
+			return reconcile.Result{Requeue: true}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	return reconcile.Result{Requeue: true}, nil
+}
+
+// IsReady checks if status of condition passed as conditionPath is true or not
+func (c *Controller) IsReady(conditionPath string, nodeClass *unstructured.Unstructured) (bool, error) {
```

Comment:
Is there anything upstream that already exists upstream for this? If not, is this something that should potentially live in [`operatorpkg`](https://github.com/awslabs/operatorpkg)